Stripe implementation

alextravin opened the pull request:
Stripe implementation. We also need to change ai-client-html. Will create pull request now

alextravin committed:
Stripe implementation

Stripe implementation

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php:
```diff
@@ -711,11 +737,17 @@ protected function getData( \Aimeos\MShop\Order\Item\Base\Iface $base, $orderid,
+		$config = $this->getServiceItem()->getConfig();
+
+		if($this->getValue( 'apiKey', false ) != false){
+			$config['apiKey'] = $this->getValue( 'apiKey', false );
```
Why do you use an if-condition here? The value of "apiKey" would be passed to initialize() nevertheless if available as long as it has no "stripe." prefix (which is OK in config settings).

alextravin replied:
Because of I added it to parent OmniPay class. So we add apiKey to an array only if child payment provider need it. Okay I can add getProvider() to child Stripe class and do it there, without if-condition

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php:
```diff
@@ -217,6 +217,20 @@ class OmniPay
 			'public' => false,
 		),
+
+		'paymenttoken' => array(
```
You should move this to the Stripe provider if it's specific for Stripe

alextravin replied:
Sorry, my bad. Done

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php:
```diff
@@ -700,6 +721,11 @@ protected function getData( \Aimeos\MShop\Order\Item\Base\Iface $base, $orderid,
 		}
 
+
+		if( $this->getValue( 'token', false ) && isset($params['paymenttoken']) ){
```
This seems to be Stripe specific too

alextravin replied:
I thought if we will have more payments systems with tokens, we can keep this code here, so it's checking if child class have tokens. Okay I will move it to child Stripe class

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -32,21 +126,309 @@ protected function getConfigPrefix()
+	{
+		switch ($key) {
+			case 'type':
+				return 'Stripe';
```
If you overwrite getProvider() you don't need that

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -32,21 +126,309 @@ protected function getConfigPrefix()
+				return 'Stripe';
+			case 'onsite':
+				return true;
+			case 'token':
```
What is "token" for? Isn't it the specific token for the payment of the customer?

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -32,21 +126,309 @@ protected function getConfigPrefix()
+	 */
+	public function getConfigFE(\Aimeos\MShop\Order\Item\Base\Iface $basket)
+	{
+		return [];
```
This method isn't used but should be so it renders the required fields (CC number, expiry, cvv) in the frontend. Then, they can be styled in the frontend and don't need to be passed as part of the HTML string in the HTML client

alextravin replied:
Sorry, a little bit misunderstanding. Is it good now?

aimeos replied:
What @aimeoscom probably means is that you should add the definition of the input fields for card-cardNumber, card-cardExpiry and card-cardCVC you currently have in the HTML to `$this->feConfig` instead. Then, they will be rendered in the HTML checkout process client template automatically. You only have to tell the Stripe JS code what are the new div containers and it hopefully replaces the existing input field in that div container.

alextravin replied:
If the method getConfigFE() will return some fields, they will be render not at the last step, it will be at payment step, something like that: http://joxi.ru/82QeqY5U1eJEg2
But we don't need so. I don't understand that part of Aimeos. How we can use getConfigFE() to render form at the last step only, and prevent it at payment step?

The second one problem is we can't append Stripe element to <input>. So we have to modify templates\checkout\standard\process-body-standard.php or append Stripe element to parent html tag  to `<li class="form-item" >`  (I think it's not good)

aimeoscom replied:
You are right, `getConfigFE()` is the wrong place and it must be `getPaymentForm()` instead I think.

What's the problem if we tell Stripe to use the `<li class="form-item" >` as container for its input elements? I would expect that it replaces the existing input fields that are inside.

alextravin replied:
Yes it will replace content of `<li class="form-item" >`. Have I to do so and leave at getHtmlForm() Stripe JS only?
I think sometime in the future we can give ability to user to check what Stripe elements he want to show.

aimeoscom replied:
Cool! Let's do it that way so the surrounding HTML can be changed in the template :-)

alextravin replied:
Then I have to create JS selector for the `li`
`<li class="form-item payment.cardno mandatory">`

I am trying $(".payment\.cardno"); but it does not work with an error at a console: 
>The selector you specified (.payment.cardno) applies to no DOM elements that are currently on the page. Make sure the element exists on the page before calling mount().

It's strange, because if I add an id to that `<li>`, it works fine. 
So I will modify ai-client-html\client\html\templates\checkout\standard\process-body-standard.php to add IDs too.

aimeoscom replied:
Something with a dot is no valid CSS class. Can you change the output in the HTML template and replace '.' by '-' for the class name?

alextravin replied:
So if we remove fields from GetHtml() and they will be render by $feConfig I want to add `id="payment-<?=$item->getInternalCode();?>"` to `<li>` tag. Is it good?

aimeos replied:
No, you don't need that if you fix the CSS classes by replacing dot by dash. Then your CSS selector will work

alextravin replied:
Yes I know it will work. I just fought it's not good to use '-' instead of dot because at all payment providers at Aimeos you are using dot as standard delimeter.

aimeoscom replied:
You must not change the keys in your Stripe class but mangle the output in the HTML client template:
https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L129

aimeoscom replied:
Simply add a `str_replace( '.', '-', $key );` there:
https://github.com/aimeos/ai-client-html/blob/master/client/html/templates/checkout/standard/process-body-standard.php#L129

alextravin replied:
Okay. Just don't understand why it's better than add an id.
Please also note now we have jquery JS at the body of the page. So it does not work until we put calling jquery.min.js to the top of the page. Or we need to improve Aimeos somehow to be able add special Stripe JS to the template's footer, after calling jquery. 
At my test domen I put jquery to the top of template.

alextravin replied:
Also as you can see at my JS I'm calling an AimeosPurchaseHandler object. It's for the future, if we will create card details form at the chosing provider step:  http://joxi.ru/xAeY510fYQNk0A

aimeoscom replied:
We try to avoid IDs as they need to be unique over the page and CSS classes don't have to. This is especially a concern when things can be dynamically extended by extensions.

Can you use `document.getElementsByClassName("...");` instead of jQuery so we don't need it?

I would suggest to remove the functionality of AimeosPurchaseHandler for the moment because it doesn't seem to be generic enough to be included into the Aimeos admin.js. You can create another PR where we can discuss how to integrate Stripe into the payment section of the checkout.

aimeoscom replied:
If the code gets to bloated or complicated, you can also use document.getElementById() instead. It's then worth the disadvantages of IDs

alextravin replied:
Please take a look at new commit. I've removed Jquery and removed AimeosPurchaseHandler. Now it's just pure JS and we didn't add any new ID to the html.

alextravin replied:
Except button's ID. We have added id="payment-button" to Pay Now button already

alextravin replied:
Please note one more thing. We have added `<script src="https://js.stripe.com/v3/"></script>`
just to checkout page, by getHtmlForm() method of form helper. But according to Stripe documentation it's recommended to place this including JS to the each one page to prevent fraud. 
I don't know how to do here. 
I still think we need some "hook" to place any necessary code to the footer or header.

aimeoscom replied:
I would ignore that recommendation for the moment because it doesn't make sense to have that file at each page. It's possible to place that in the HTML head if you add it to the process-header-standard template. But you would have to create a decorator for this that adds that part to the view and the template has to render that information afterwards.

alextravin replied:
Yes, I don't insist on it, just notice. Or maybe we can explain it as description of provider to users who will use it.

alextravin committed:
Stripe implementation

Stripe implementation

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php:
```diff
@@ -545,6 +546,13 @@ public function updatePush( \Psr\Http\Message\ServerRequestInterface $request, \
 	 */
 	public function updateSync( \Psr\Http\Message\ServerRequestInterface $request, \Aimeos\MShop\Order\Item\Iface $order )
 	{
+		$base = $this->getOrderBase( $order->getBaseId() );
```
These lines seems to be duplicate

alextravin replied:
Hmm... Didn't touch the updateSync() method. Don't see duplicate lines. What should I do now?

aimeos replied:
Can you make sure the implementation in your fork and PR is the same as in the current master branch so the unit tests doesn't fail any more:
https://github.com/aimeoscom/ai-payments/blob/master/lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php#L546

alextravin replied:
now we have deleted all new code from Omnipay.php, so it will stay original like at master branch I think

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/OmniPay.php:
```diff
@@ -919,7 +927,7 @@ protected function processOrder( \Aimeos\MShop\Order\Item\Iface $order, array $p
 
 	/**
-	 * Adds the transation reference to the order service attributes.
+	 * Addes the transation reference to the order service attributes.
```
"Adds" was correct :-)

alextravin replied:
I didn't touch it. Strange :-)

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +129,276 @@ protected function getConfigPrefix()
+	protected function getProvider()
+	{
+		$config = $this->getServiceItem()->getConfig();
+		$config['apiKey'] = $this->getValue('apiKey','');
```
This is unnecessary as 'apiKey' is already set in $config

alextravin replied:
I did not clearly understanding. Why it is not necessary? If we do not pass it there, we will get "Invalid API Key provided:" from the Stripe. We have that key in the $config, but as $config['stripe.apiKey'];
Do you mean we have to write
$config['apiKey'] = $config['stripe.apiKey'];
instead of 
$config['apiKey'] = $this->getValue('apiKey','');
?

alextravin replied:
You did not provide an API key. You need to provide your API key in the Authorization header, using Bearer auth (e.g. 'Authorization: Bearer YOUR_SECRET_KEY'). See https://stripe.com/docs/api#authentication for details, or we can help at https://support.stripe.com/.

aimeos replied:
OK, got it. Then it's fine.

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -21,6 +22,101 @@ class Stripe
 {
+
+	private $beConfig = array(
+		'type' => array(
```
'type' and 'onsite' are not required because they are fixed for Stripe

alextravin replied:
Cool. Removed

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -21,6 +22,101 @@ class Stripe
+			'default'=> '0',
+			'required'=> false,
+		),
+		'apiKey' => array(
```
Are 'apiKey' and 'publishable_key' named in the Stripe documentation exactly in this spelling

alextravin replied:
For the first 'apiKey' I gave that name because this is how we have to call it when pushing to Omnipay. 
For the second, at Stripe documentation it's called as "Publishable key" http://joxi.ru/12MepDjU4DoEO2
And at the Stripe code examples it don't have specific name, they just insert your key itself

aimeos replied:
OK, in this case can you rename it to "publishableKey" so it's consistent to "apiKey"?

alextravin replied:
Sure

alextravin committed:
Stripe implementation

Stripe implementation

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,276 @@ protected function getConfigPrefix()
+		//return $this->checkConfig($this->feConfig, $attributes);
+	}
+
+	public function getHTMLform()
```
Can you name the method according to the camel case standard, i.e. getHtmlForm()?

alextravin replied:
Done

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,276 @@ protected function getConfigPrefix()
+		$list = [];
+
+		if( $prefix !== 'omnipay' ) {
+			unset( $config['type'], $config['onsite'] );
```
Seems they are gone, at least they are not listed in the backend config any more

alextravin replied:
Removed

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,276 @@ protected function getConfigPrefix()
+		$config = $this->beConfig;
+
+		if( $prefix !== 'omnipay' ) {
+			unset( $config['type'], $config['onsite'] );
```
These seem to be gone too

alextravin replied:
Removed

alextravin committed:
Restore original Omnipay.php, rename publishable_key to publishableKey, rename getHTMLform to getHtmlForm

Restore original Omnipay.php, rename publishable_key to publishableKey,
rename getHTMLform to getHtmlForm, remove unnecessary methods

aimeoscom commented:
Did you copy the OmniPay.php from master to your fork? It still looks wrong because GitHub shows changes in the file and there should be none.

alextravin committed:
Update OmniPay.php

alextravin commented:
Yes I did so. And did it now again. 
What can I do to fix it up?

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -21,6 +22,83 @@ class Stripe
+		),
+	);
+
+	protected $feConfig = array(
```
Do we need this at all if we use the payment form at the process step?

alextravin replied:
Yes we need it. Without it **process-body-standard.php** is not rendering fields.

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,225 @@ protected function getConfigPrefix()
+		return '<script src="https://js.stripe.com/v3/"></script>
+		<script type="text/javascript">
+    	$(document).ready(function () {
+        var stripe = Stripe("'.$this->getConfigValue( array( $this->getConfigPrefix() . '.publishableKey' ), '' ).'");
```
You can use `getConfigValue( 'stripe.publishableKey' )` instead

alextravin replied:
Yes I know. Okay will do so :-)

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,225 @@ protected function getConfigPrefix()
+		return '<script src="https://js.stripe.com/v3/"></script>
+		<script type="text/javascript">
+    	$(document).ready(function () {
+        var stripe = Stripe("'.$this->getConfigValue( array( $this->getConfigPrefix() . '.publishableKey' ), '' ).'");
```
Can you put everything in an JS object so it doesn't pollute the global JS name space?
Furthermore you should remove all code that is inactive and all debug statements.

alextravin replied:
Okay, I will :-)

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +111,225 @@ protected function getConfigPrefix()
+            //return null;
+            
+            
+            AimeosPurchaseHandler.submitPurchaseForm();
```
Where is this code from?

alextravin committed:
Delete unnecessary html fields from getHtmlForm

alextravin committed:
Merge branch 'master' of https://github.com/alextravin/ai-payments

alextravin committed:
JS to an object, new selectors. Fields are not required anymore, because the form is not submitting if we set the fields required and they are replaced by Stripe

aimeos commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -21,6 +22,99 @@ class Stripe
+
+	protected $feConfig = array(
+
+		'paymenttoken' => array(
```
What about the payment token? Is that the token we need to store for executing the payment? When exactly is the payment executed?

alextravin replied:
When we send card details to Stripe, we take token as Stripe's answer. Then I put the token at this hidden field 'paymenttoken', and this is only one value we have to send to Stripe later via Omnipay. So this field is necessary.

alextravin replied:
Payment will be executed when we'll send token to Stripe by Omnipay.

aimeoscom commented on lib/custom/src/MShop/Service/Provider/Payment/Stripe.php:
```diff
@@ -33,20 +127,184 @@ protected function getConfigPrefix()
+
+	<!-- Used to display Element errors -->
+	<div id="card-errors" role="alert"></div>
+	<input type="hidden" id="CurrentPaymentMethod" value="Stripe" />';
```
What is this input field for?

alextravin committed:
Pure JS instead of Jquery

aimeoscom commented:
Looks good now :-)
Can you please adapt the unit tests that fails at the moment and the coding style?

alextravin committed:
Tiny changes

alextravin commented:
Sure, I will. Trying to understand phing at the moment :-)

alextravin committed:
Deleted 2 tests because of they are not actual for current version of Stripe provider
